association-backend/src/webhooks/handlers/pretix/pretix_event_order_paid.py

```diff
@@ -0,0 +1,161 @@
+from src.webhooks.exceptions import NoUserFoundWithEmail
+
+from .api import PretixAPI
+
+logger = logging.getLogger(__file__)
+
+GET_USER_ID_BY_EMAIL = """query UserIdByEmail($email: String!) {
+    userByEmail(email: $email) {
+        id
+    }
+}"""
+
+
+async def pretix_event_order_paid(payload):
+    action = payload["action"]
+    organizer = payload["organizer"]
+    event = payload["event"]
+    order_code = payload["code"]
+
+    pretix_api = PretixAPI(organizer, event)
+
+    order_data = pretix_api.get_order_data(order_code)
+    categories = {
+        result["id"]: result for result in pretix_api.get_categories()["results"]
+    }
+
+    order_positions = order_data["positions"]
+    membership_positions = []
+    for position in order_positions:
+        item_data = pretix_api.get_item_data(position["item"])
+        category = categories.get(item_data["category"], None)
+
+        assert (
+            category
+        ), f"order_code={order_code} item={item_data['id']} does not have a valid category={item_data['category']}"
+
+        # Items that are part of a category named "Association"
+        # are considerated special items that should subscribe the user to the association
+        if category["internal_name"] != "Association":
+            continue
+
+        membership_positions.append(position)
+
+    if not membership_positions:
+        logger.info(
+            "No membership positions for order_code=%s so nothing to do", order_code
+        )
+        return
+
+    if len(membership_positions) > 1:
+        raise ValueError(
+            f"Multiple positions found in order_code={order_code} that subscribe the user to the association. This is not supported."
+        )
+
+    user_email = order_data["email"]
+    client = ServiceClient(
+        url=f"{USERS_SERVICE_URL}/internal-api",
+        service_name="users-backend",
+        caller="association-backend",
+        jwt_secret=str(SERVICE_TO_SERVICE_SECRET),
+    )
+    result = await client.execute(GET_USER_ID_BY_EMAIL, {"email": user_email})
+    data = result.data
+
+    if not data["userByEmail"]:
+        raise NoUserFoundWithEmail(
+            f"No user found with the email of order_code={order_code}"
+        )
+
+    user_id = int(data["userByEmail"]["id"])
+    repository = AssociationMembershipRepository()
+
+    idempotency_key = PretixPayment.generate_idempotency_key(
+        organizer, event, order_code
+    )
+    if await repository.is_payment_already_processed(idempotency_key):
+        logger.info(
+            "Ignoring action=%s (organizer=%s event=%s) from Pretix because we already processed "
+            "the payment with key=%s from order_code=%s",
+            action,
+            organizer,
+            event,
+            idempotency_key,
+            order_code,
+        )
+        return
+
+    subscription = await repository.get_user_subscription(user_id)
```

Comment:
What about the case when the user already has a subscription but it's not active?

Reply:
The subscription will get marked as active again

Reply:
Ok, but here we will get `None` or something in `subscription` var? 
Because if we got an object with `status.CANCELED` for example, on line 111 the check will fail, isn't it?

Reply:
If the subscription is canceled it will still be found, just with status CANCELED

>  Because if we got an object with status.CANCELED for example, on line 111 the check will fail, isn't it?

Maybe we are looking at a different code?

If no subscription is found, a new one is created: https://github.com/pythonitalia/pycon/pull/2557/files/12f5633268f4b01a2b92533cdb240de8bdb7b2ca#diff-06f612155123ab4a22841f0ccbef0d6aa5bd18fc876bba451f23838e6e240dd4R101 This will happen when the user never subscribed to the association, a subscription gets created on the fly.

If a subscription is found, but is active, we raise an error: https://github.com/pythonitalia/pycon/pull/2557/files/12f5633268f4b01a2b92533cdb240de8bdb7b2ca#diff-06f612155123ab4a22841f0ccbef0d6aa5bd18fc876bba451f23838e6e240dd4R104 If that happens, we would have to refund them the 10EUR manually.

Let me know if I am looking at the wrong place, I am not 100% sure I understand what you mean here #1431

Reply:
Oh right, you don't have to recreate the object subscription again... ok then